### PR TITLE
added deit onnx config

### PR DESCRIPTION
**Pull request by 0xrushi**
# What does this PR do?

Added DeiT OnnxConfig to make this model available for conversion

@ChainYo 

https://github.com/huggingface/transformers/issues/16308

Fixes # (issue)

## Before submitting
- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [ ] Did you read the [contributor guideline](https://github.com/huggingface/transformers/blob/main/CONTRIBUTING.md#start-contributing-pull-requests),
      Pull Request section?
- [x] Was this discussed/approved via a Github issue or the [forum](https://discuss.huggingface.co/)? Please add a link
      to it if that's the case.
- [x] Did you make sure to update the documentation with your changes? Here are the
      [documentation guidelines](https://github.com/huggingface/transformers/tree/main/docs), and
      [here are tips on formatting docstrings](https://github.com/huggingface/transformers/tree/main/docs#writing-source-documentation).
- [ ] Did you write any new necessary tests?

## Who can review?

Anyone in the community is free to review the PR once the tests have passed. Feel free to tag
members/contributors who may be interested in your PR.

### CODE REVIEW

File: src/transformers/onnx/features.py [hunk 1]
```diff
@@ -286,6 +287,7 @@ class FeaturesManager:
         ),
         "vit": supported_features_mapping("default", "image-classification", onnx_config_cls=ViTOnnxConfig),
         "beit": supported_features_mapping("default", "image-classification", onnx_config_cls=BeitOnnxConfig),
+        "deit": supported_features_mapping("default", "image-classification", onnx_config_cls=DeiTOnnxConfig),
```
Comment (ChainYo) [inline src/transformers/onnx/features.py:290]: I saw that DeiT also have a `DeiTForMaskedImageModeling` implementation, is it a feature also available in onnx ?

**Comment by ChainYo** [comment]
> # What does this PR do?
> Added DeiT OnnxConfig to make this model available for conversion
> 
> @ChainYo

> ## Who can review?
> Anyone in the community is free to review the PR once the tests have passed. Feel free to tag members/contributors who may be interested in your PR.


Also pinging @LysandreJik and @NielsRogge because he did the implementation of DeiT.

File: src/transformers/onnx/features.py (hunk 1 above)
Comment (rushic24) [reply src/transformers/onnx/features.py:290]: BeiT also has BeiTForMaskedImageModeling, but there isn't anything for that in features.py

Comment (lewtun) [reply src/transformers/onnx/features.py:290]: You can add something like 

```
"masked-im": AutoModelForMaskedImageModeling
```

to the `_TASKS_TO_AUTOMODELS` attribute of the `FeaturesManager`. I don't think we currently support this feature for TensorFlow, so no need to include this in `__TASKS_TO_TF_AUTOMODELS`.

Feel free to then include this feature as part of `vit`, `deit`, and `beit`

**Review by lewtun — COMMENTED** [review]
Thanks for adding DEIT to the ONNX exporter @rushic24 ! Overall this looks great and I think it would be neat if we can also include the masked image modelling feature :)

**Comment by rushic24** [comment]
Thanks, I just added those.

I'm trying to add a README  to [ONNXConfigForAll](https://huggingface.co/OWG),

How does onnx work in ViT? 
I tried the code below, but its not working

```
from transformers import ViTFeatureExtractor, ViTModel
import torch
from datasets import load_dataset
from onnxruntime import InferenceSession


dataset = load_dataset("huggingface/cats-image")
image = dataset["test"]["image"][0]

feature_extractor = ViTFeatureExtractor.from_pretrained("google/vit-base-patch16-224-in21k")
model = ViTModel.from_pretrained("google/vit-base-patch16-224-in21k")

inputs = feature_extractor(image, return_tensors="pt")

session = InferenceSession("onnx/model.onnx")

# ONNX Runtime expects NumPy arrays as input
outputs = session.run(output_names=["last_hidden_state"], input_feed=list(inputs))
```

**Comment by ChainYo** [comment]
> How does onnx work in ViT?
> I tried the code below, but its not working

Could you check on [netron.app](https://netron.app) what are the inputs of the onnx converted model and then check if the `input_feed` is correct. Btw it should be a `Dict` not a `List`.

**Comment by rushic24** [comment]
> > How does onnx work in ViT?
> > I tried the code below, but its not working
> 
> Could you check on [netron.app](https://netron.app) what are the inputs of the onnx converted model and then check if the `input_feed` is correct. Btw it should be a `Dict` not a `List`.

And inputs itself is a dictionary
image


The input as seen from netron is pixel_values, I did try dict(inputs), inputs, list(inputs)

image

**Comment by rushic24** [comment]
I just figured it out. You can find it here [https://huggingface.co/OWG](https://huggingface.co/OWG) :-)

**Comment by ChainYo** [comment]
> I just figured it out. You can find it here https://huggingface.co/OWG/DeiT :-)

It seems really good! :hugs:

**Review by lewtun — APPROVED** [review]
Thanks for iterating on this @rushic24 and great job on pushing a model to the Hub 😍 !

This LGTM so gently pinging @sgugger for final approval

File: src/transformers/onnx/features.py [hunk 2]
```diff
@@ -102,6 +104,7 @@ class FeaturesManager:
             "multiple-choice": AutoModelForMultipleChoice,
             "question-answering": AutoModelForQuestionAnswering,
             "image-classification": AutoModelForImageClassification,
+            "masked-im": AutoModelForMaskedImageModeling,
```
Comment (lewtun) [inline src/transformers/onnx/features.py:107]: For reviewers: we picked `masked-im` as the name to match the pre-existing `masked-lm` used for language modeling

**Review by sgugger — APPROVED** [review]
Thanks for the PR. It all looks good but there is a `datasets` submodule added in this PR that should be removed before it can be merged.

**Review by chainyo — APPROVED** [review]

**Comment by lewtun** [comment]
Thanks for the fix!
